File: Data Structures (Java & C++)/Project 4/Project 4 C++/BinarySearchTree.cpp
//
//  BinarySearchTree.cpp
//  Project 4 C++
//
//  Created by Alexandru Gutu on 11/14/14.
//  Copyright (c) 2014 Alexandru Gutu. All rights reserved.
//

#include "BinarySearchTree.h"

// BST Constructor with significantly large dummy node
BinarySearchTree::BinarySearchTree()
{
	root = new Node(999999999);
}

// Returns the root of the BST
Node *BinarySearchTree::getRoot()
{
	return root;
}

void BinarySearchTree::insert(Node *current, int data)
{
	bool done = false;
	
	// Case 1: Data matches
	if (current->getData() == data)
	{
		cout << "Unable to insert, duplicate data [" << data << "]" << endl;
		cout << endl;
	}
	
	// Case 2: Incoming data is less than current node's data
	else if (current->getData() > data)
	{
		if (current->getLeft() == nullptr)
		{
			current->setLeft(new Node(data));
			done = true;
		}
		else
		{
			insert(current->getLeft(), data);
		}
	}
	
	// Case 3: Incomign data is larger than current node's data
	else if (current->getData() < data)
	{
		if (current->getRight() == nullptr)
		{
			current->setRight(new Node(data));
			done = true;
		}
		else
		{
			insert(current->getRight(), data);
		}
	}
}

// Search for data
void BinarySearchTree::search(int data)
{
	Node *current = root->getLeft();
	
	bool isFound = false;
	
	cout << endl;
	cout << "Search steps for " << data << endl;
	
	while (current != nullptr)
	{
		current->printData();
		
		if (current->getData() == data)
		{
			cout << "Node of value " << data << " has been found." << endl;
			cout << endl;
			isFound = true;
			break;
		}
		else
		{
			if (current->getData() > data)
			{
				current = current->getLeft();
			}
			else
			{
				current = current->getRight();
			}
		}
	}
	
	if (isFound == false)
	{
		cout << "The node of value " << data << " has not been found" << endl;
		cout << endl;
	}
}

// Printing in Pre, In, and Post order (all 3 are kept for archiving sake)
void BinarySearchTree::printFormat(string format)
{
	Node *current = root->getLeft();
	
	if (format == "pre")
	{
		preOrder(current);
	}
	else if (format == "in")
	{
		inOrder(current);
	}
	else if (format == "post")
	{
		postOrder(current);
	}
	else
	{
		cout << "Unknown operation modifier! Please check input file and try again" << endl;
	}
}

void BinarySearchTree::preOrder(Node *current)
{
	if (current == nullptr)
	{
		return;
	}
	
	current->printData();
	preOrder(current->getLeft());
	preOrder(current->getRight());
}

void BinarySearchTree::inOrder(Node *current)
{
	if (current == nullptr)
	{
		return;
	}
	
	preOrder(current->getLeft());
	current->printData();
	preOrder(current->getRight());
}

void BinarySearchTree::postOrder(Node *current)
{
	if (current == nullptr)
	{
		return;
	}
	
	preOrder(current->getLeft());
	preOrder(current->getRight());
	current->printData();
}

File: Data Structures (Java & C++)/Project 4/Project 4 Java/BinarySearchTree.java
/**
 * Project Name: Project 4 Java
 * Created by Alexandru Gutu on 11/1/2014.
 * Copyright (c) 2014 Alexandru Gutu. All rights reserved.
 */

public class BinarySearchTree
	{
		private Node root;

		// BST Constructor with significantly large dummy node
		public BinarySearchTree()
			{
				root = new Node(999999999);
			}

		// Returns the root of the BST
		public Node getRoot()
			{
				return root;
			}

		public void insert(Node current, int data)
			{
			// Case 1: Data matches
			if (current.getData() == data)
				{
					System.out.println("Unable to insert, duplicate data [" + data + "]");
					System.out.println();
				}

			// Case 2: Incoming data is less than current node's data
			else if (current.getData() > data)
				{
					if (current.getLeft() == null)
						{
							current.setLeft(new Node(data));
							return;
						}
					else
						{
							insert(current.getLeft(), data);
						}
				}

			// Case 3: Incoming data is more than current node's data
			else if (current.getData() < data)
				{
					if (current.getRight() == null)
						{
							current.setRight(new Node(data));
							return;
						}
					else
						{
							insert(current.getRight(), data);
						}
				}
			}

		// Search for data
		public void search(int data)
			{
				Node current = root.getLeft();

				boolean isFound = false;

				System.out.println();
				System.out.println("Search Steps for " + data);

				while (current != null)
					{
						current.printData();

						if (current.getData() == data)
							{
								System.out.println("Node of value " + data + " has been found.");
								System.out.println();
								isFound = true;
								break;
							}
						else
							{
								if (current.getData() > data)
									{
										current = current.getLeft();
									}
								else
									{
										current = current.getRight();
									}
							}
					}

				if (!isFound)
					{
						System.out.println("The node of value " + data + " has not been found");
						System.out.println();
					}
			}

		// Printing in Pre, In, Post order (All 3 are kept for archiving sake)
		public void printFormat(String format)
			{

				Node current = root.getLeft();

				if (format.equals("pre"))
					{
						preOrder(current);
					}
				else if (format.equals("in"))
					{
						inOrder(current);
					}
				else if (format.equals("post"))
					{
						postOrder(current);
					}
				else
					{
						System.out.println("Unknown operation modifier! Please check input file and try again");
					}
			}

		public void preOrder(Node current)
			{
				if (current == null)
					{
						return;
					}

				current.printData();

				preOrder(current.getLeft());
				preOrder(current.getRight());
			}

		public void inOrder(Node current)
			{
				if (current == null)
					{
						return;
					}

				preOrder(current.getLeft());
				current.printData();
				preOrder(current.getRight());
			}

		public void postOrder(Node current)
			{
				if (current == null)
					{
						return;
					}

				postOrder(current.getLeft());
				postOrder(current.getRight());
				current.printData();
			}
	}
